x/oracle/keeper/tally_test.go

```diff
@@ -0,0 +1,201 @@
+			// Tally 2 votes
+			// rewardBand: 1.000000000000000000
+			// weightedMedian: 1.000000000000000000
+			// standardDeviation: 0.353553390593273762
+			// rewardSpread: 0.500000000000000000
+
+			Ballot: types.ExchangeRateBallot{
+				types.NewVoteForTally(sdk.MustNewDecFromStr("1.0"), "AAA", val(1), 10000),
+				types.NewVoteForTally(sdk.MustNewDecFromStr("1.5"), "AAA", val(2), 10000),
+			},
+			RewardBand: sdk.MustNewDecFromStr("1.0"),
+			ValidatorClaimMap: map[string]types.Claim{
+				val(1).String(): types.NewClaim(10000, 0, 0, val(1)),
+				val(2).String(): types.NewClaim(10000, 0, 0, val(2)),
+			},
+			ExpectedWeightedMedian: sdk.MustNewDecFromStr("1.0"),
+			ExpectedValidatorClaimMap: map[string]types.Claim{
+				val(1).String(): types.NewClaim(10000, 10000, 1, val(1)),
+				val(2).String(): types.NewClaim(10000, 10000, 1, val(2)),
+			},
+		},
+		{
+			Description: "basic test for >50%% attack",
+			// Tally 5 votes
+			// rewardBand: 1.000000000000000000
+			// weightedMedian: 999.000000000000000000
+			// standardDeviation: 892.638336617916046807
+			// rewardSpread: 892.638336617916046807
+
+			Ballot: types.ExchangeRateBallot{
+				types.NewVoteForTally(sdk.MustNewDecFromStr("1.0"), "AAA", val(1), 10000),
+				types.NewVoteForTally(sdk.MustNewDecFromStr("1.0"), "AAA", val(2), 10000),
+				types.NewVoteForTally(sdk.MustNewDecFromStr("1.0"), "AAA", val(3), 10000),
+				types.NewVoteForTally(sdk.MustNewDecFromStr("1.0"), "AAA", val(4), 10000),
+				types.NewVoteForTally(sdk.MustNewDecFromStr("999.0"), "AAA", val(5), 40002),
+			},
+			RewardBand: sdk.MustNewDecFromStr("1.0"),
+			ValidatorClaimMap: map[string]types.Claim{
+				val(1).String(): types.NewClaim(10000, 0, 0, val(1)),
+				val(2).String(): types.NewClaim(10000, 0, 0, val(2)),
+				val(3).String(): types.NewClaim(10000, 0, 0, val(3)),
+				val(4).String(): types.NewClaim(10000, 0, 0, val(4)),
+				val(5).String(): types.NewClaim(40002, 0, 0, val(5)),
+			},
+			ExpectedWeightedMedian: sdk.MustNewDecFromStr("999.0"),
+			ExpectedValidatorClaimMap: map[string]types.Claim{
+				val(1).String(): types.NewClaim(10000, 0, 0, val(1)),
+				val(2).String(): types.NewClaim(10000, 0, 0, val(2)),
+				val(3).String(): types.NewClaim(10000, 0, 0, val(3)),
+				val(4).String(): types.NewClaim(10000, 0, 0, val(4)),
+				val(5).String(): types.NewClaim(40002, 40002, 1, val(5)),
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		s.T().Log("TestTally Case:", testCase.Description)
+
+		resultValidatorClaimMap := copyValidatorClaimMap(testCase.ValidatorClaimMap)
+		weightedMedian, err := keeper.Tally(
+			testCase.Ballot,
+			testCase.RewardBand,
+			resultValidatorClaimMap,
+		)
+
+		s.Require().NoError(err)
+		s.Require().EqualValues(testCase.ExpectedWeightedMedian, weightedMedian)
+		s.Require().EqualValues(testCase.ExpectedValidatorClaimMap, resultValidatorClaimMap)
+	}
+}
+
+type tallyTestCase struct {
+	Description string
```

Comment:
The struct itself is private to be used in the testing only, but fields are public interface for test case, public fields are expected to be set by test case writer. As opposed to test case runner logic that may benefit from private fields.

For example, private fields here might be some cached/parsed values or even `mux sync.Mutex` in case it is used in some async processing. Certainly overthinking for this particular test, but in general there is nothing wrong with public fields in a private struct.

Also, if you want to quickly dump the struct as JSON, it will be impossible to do if fields were private.